libraries/ms-drivers/test/test_ads1259_adc.c

```diff
@@ -0,0 +1,147 @@
+#include "ads1259_adc.h"
+#include "delay.h"
+#include "gpio.h"
+#include "interrupt.h"
+#include "log.h"
+#include "math.h"
+#include "soft_timer.h"
+#include "test_helpers.h"
+#include "unity.h"
+
+#define TEST_AD1259_SPI_PORT SPI_PORT_2
+#define NUM_CONFIG_REGISTERS 3
+#define TEST_CHK_SUM_FLAG_BIT 0x80
+#define TEST_DRDY_BIT 0x80
+#define TEST_DATA_SETTLING_TIME_MS 17
+
+#define TEST_BAUDRATE 60000
+#define TEST_MOSI_PIN \
+  { .port = GPIO_PORT_B, 15 }
+#define TEST_MISO_PIN \
+  { .port = GPIO_PORT_B, 14 }
+#define TEST_SCLK_PIN \
+  { .port = GPIO_PORT_B, 13 }
+#define TEST_CS_PIN \
+  { .port = GPIO_PORT_B, 12 }
+
+typedef struct Ads1259MockRegisters {
+  uint8_t CONFIG_0;
+  uint8_t CONFIG_1;
+  uint8_t CONFIG_2;
+  uint8_t OFC_0;
+  uint8_t OFC_1;
+  uint8_t OFC_2;
+  uint8_t FSC_0;
+  uint8_t FSC_1;
+  uint8_t FSC_2;
+} Ads1259MockRegisters;
+
+static uint8_t register_lookup[NUM_CONFIG_REGISTERS] = {
+  (ADS1259_SPI_TIMEOUT_ENABLE),
+  (ADS1259_OUT_OF_RANGE_FLAG_ENABLE | ADS1259_CHECK_SUM_ENABLE),
+  (ADS1259_CONVERSION_CONTROL_MODE_PULSE | ADS1259_DATA_RATE_SPS),
+};
+
+static Ads1259Storage storage;
+static Ads1259MockRegisters registers;
+
+StatusCode TEST_MOCK(spi_exchange)(SpiPort spi, uint8_t *tx_data, size_t tx_len, uint8_t *rx_data,
+                                   size_t rx_len) {
+  TEST_ASSERT_EQUAL(spi, TEST_AD1259_SPI_PORT);
+  uint8_t cmd = tx_data[0];
+  LOG_DEBUG("%x\n", cmd);
+  if (cmd & ADS1259_READ_REGISTER) {
+    if (cmd == 0x20) rx_data[0] = registers.CONFIG_0;
+    if (cmd == 0x21) rx_data[0] = registers.CONFIG_1;
+    if (cmd == 0x22) rx_data[0] = registers.CONFIG_2;
+    return STATUS_CODE_OK;
+  }
+  switch (cmd) {
+    // Commands used in ads1259_init()
+    case ADS1259_STOP_READ_DATA_CONTINUOUS:
+      LOG_DEBUG("STOP DATA CMD SENT\n");
+      return STATUS_CODE_OK;
+    case ADS1259_RESET:
+      LOG_DEBUG("RESET CMD SENT\n");
+      return STATUS_CODE_OK;
+    case (ADS1259_WRITE_REGISTER | ADS1259_ADDRESS_CONFIG0):
+      LOG_DEBUG("WRITE_REG CMD SENT\n");
+      TEST_ASSERT_TRUE(tx_len == 5 && rx_len == 0);
+      TEST_ASSERT_TRUE(tx_data[1] == 3);
+      registers.CONFIG_0 = tx_data[2];
+      registers.CONFIG_1 = tx_data[3];
+      registers.CONFIG_2 = tx_data[4];
+      break;
+    case ADS1259_OFFSET_CALIBRATION:
+      LOG_DEBUG("OFFSET CALIBRATION SENT\n");
+      registers.OFC_0 = 0;
+      registers.OFC_1 = 0;
+      registers.OFC_2 = 0;
+      break;
+    case ADS1259_GAIN_CALIBRATION:
+      LOG_DEBUG("GAIN CALIBRATION SENT\n");
+      registers.FSC_0 = 0;
+      registers.FSC_1 = 0;
+      registers.FSC_2 = 0;
+      break;
+    // Start of Ads1259_get_conversion_data()
+    case ADS1259_START_CONV:
+      LOG_DEBUG("CONVERSIONS STARTED \n");
+      break;
+    case ADS1259_READ_DATA_BY_OPCODE:
+      LOG_DEBUG("SETTING CONVERSION DATA\n");
+      rx_data[0] = 0xFF;
+      rx_data[1] = 0xFF;
+      rx_data[2] = 0xFF;
```

Comment:
rather than just returning `0xFF` here, what you can do is have a global static `uint8_t[4] s_rx_data` that you can set in tests then set the `rx_data` here to that static one. That way, you can set it to return various things in tests, then assert that your driver acts correctly according to what it gets in the rx.